### PR TITLE
3 implement asynchronous client layer

### LINKED ISSUE #3: 0.6.3
# Client Layer

should provide immediate answer to: 
(fail fast)

1. What is Requestor's Authentication?
2. What is that Requestor Authorized to do?
A. Access this endpoint?
B. Exclude any policies from the evaluation decision?
3. Has Consumer granted ControlBroker required read access to provided Input S3 objects?

### Generic

- [x] generic consumer Client SFN endpoint returns IsCompliant decision synchronous response

### For each Consumer:
- [ ] 1. local dev
- [ ] 2. IaC Pipeline. tracked [here](https://github.com/VerticalRelevance/control-broker-codepipeline-example).
- [ ] 3. Config-Detective

# Access Control Model

## Owned by Control Broker

APIGW endpoint
DetermineRequestValidity (a.k.a. InvokedByAigw) lambda
Evaluation Engine
PaC Policies
ResultsReport Bucket

### Access Control to ResultsReport Bucket

APIGW returns S3Uri to ResultsReport - key is UUID
consumer that has that UUID and is within the org can retry and getObject as they see fit
- [ ] implement same org bucket policy

## Owned by Consumer

ControlBrokerInputs Bucket

### Access Control to ControlBrokerInput Bucket

ControlBroker makes known [how?] the ARNs of all ControlBroker Readers needing read access to ControlBrokerInput S3 objects.

Consumer grants that read access.

# Appendix:

A. tradeoff:

SFN Standard - no pay to wait - no sync response

SFN Express - yes sync response - but cannot use .waitForTaskToken to wait for things

B. principal:

Anticipate that Consumers hitting the CB APIGW endpoint will handle the response and any retries differently. Do not be too prescriptive.

### PR DESCRIPTION
**Pull request by cschneider-vertical-relevance**
NB Config Consumer factored out to [Issue #5](https://github.com/VerticalRelevance/control-broker/issues/5)

### CODE REVIEW

File: stacks/client_stack.py [hunk 1]
```diff
@@ -140,8 +138,10 @@ def apigw(self):
             # authorizer=authorizer_iam
         )
         
-        self.apigw_full_invoke_url = f'{self.http_api.url}{self.path}'
+        self.apigw_full_invoke_url = f'{self.http_api.url[:-1]}{self.path}' # remove duplicate slash
```
Comment (eppeters) [inline stacks/client_stack.py:141]: A more robust way of doing this is the following:

```
In [1]: from os import path

In [5]: path.join("https://google.com/some", "path".strip('/'))
Out[5]: 'https://google.com/some/path'

In [6]: path.join("https://google.com/some", "///path/".strip('/'))
Out[6]: 'https://google.com/some/path'

In [7]: path.join("https://google.com/some////", "///path/".strip('/'))
Out[7]: 'https://google.com/some////path'

In [8]: path.join("https://google.com/some////".rstrip("/"), "///path/".strip('/'))
Out[8]: 'https://google.com/some/path'
```

It allows for both the base and the path to either have slashes or not, and uses only one slash in the final result. Note that I added a `.rstrip()` call in the last example, but that isn't totally necessary.

File: supplementary_files/lambdas/write_results_report/lambda_function.py [hunk 2]
```diff
@@ -58,13 +58,23 @@ def simple_pk_query(*,
         
         return items
 
-def determine_compliance(*,InfractionItems):
-    return not bool(InfractionItems)
+def determine_compliance(*,InfractionItems,AllNestedSfnsSucceeded):
```
Comment (eppeters) [inline supplementary_files/lambdas/write_results_report/lambda_function.py:61]: FYI PEP 8 defines conventions for Python code style and nearly all code checking tools, big open source projects, and generally Python projects will be expected to follow these standards.

In this case I see a violation of the following:

> Function names should be lowercase, with words separated by underscores as necessary to improve readability.
> Variable names follow the same convention as function names.

The best way I've seen to follow/learn these rules is:

https://flake8.pycqa.org/en/latest/

**Review by eppeters — CHANGES_REQUESTED** [review]
Just a few quick comments.

File: stacks/.~c9_invoke_P8ajVj.py [hunk 3]
```diff
@@ -0,0 +1,671 @@
+import os
```
[the remark this answers is not in the document]
Comment (cschneider-vertical-relevance) [reply stacks/.~c9_invoke_P8ajVj.py:1]: Added

File: README.md [hunk 4]
```diff
@@ -76,4 +99,16 @@ After running `cdk deploy`, the Control Broker will be set up.
 
 ## Next Steps
 
-Try launching one of the [Example use cases](./README.md#example-use-cases)!
\ No newline at end of file
+Try launching one of the [Example use cases](./README.md#example-use-cases)!
+
+# Setup
```
[the remark this answers is not in the document]
Comment (cschneider-vertical-relevance) [reply README.md:104]: Looks like requirements.txt has:
```
aws-cdk.aws-apigatewayv2-alpha==2.22.0a0
aws-cdk.aws-apigatewayv2-authorizers-alpha==2.22.0a0
aws-cdk.aws-apigatewayv2-integrations-alpha==2.22.0a0
aws-cdk.aws-lambda-python-alpha==2.22.0a0
```

is your env setting up properly without those additional `pip` commands?

File: supplementary_files/lambdas/write_results_report/lambda_function.py (hunk 2 above)
Comment (cschneider-vertical-relevance) [reply supplementary_files/lambdas/write_results_report/lambda_function.py]: Addressed

**Review by eppeters — APPROVED** [review]

**Pull request merged**